packages/react-native-executorch/common/rnexecutorch/models/speech_to_text/asr/ASR.cpp

```diff
@@ -0,0 +1,302 @@
+#include <random>
+
+#include "ASR.h"
+#include "executorch/extension/tensor/tensor_ptr.h"
+#include "rnexecutorch/data_processing/dsp.h"
+#include "rnexecutorch/data_processing/gzip.h"
+
+ASR::ASR(const models::BaseModel &encoder, const models::BaseModel &decoder,
+         const TokenizerModule &tokenizer)
+    : encoder(encoder), decoder(decoder), tokenizer(tokenizer),
+      startOfTranscriptionToken(
+          this->tokenizer.tokenToId("<|startoftranscript|>")),
+      endOfTranscriptionToken(this->tokenizer.tokenToId("<|endoftext|>")),
+      timestampBeginToken(this->tokenizer.tokenToId("<|0.00|>")) {}
+
+std::vector<int32_t>
+ASR::getInitialSequence(const DecodingOptions &options) const {
+  std::vector<int32_t> seq;
+  seq.push_back(this->startOfTranscriptionToken);
+
+  if (options.language.has_value()) {
+    int32_t langToken =
+        this->tokenizer.tokenToId("<|" + options.language.value() + "|>");
+    int32_t taskToken = this->tokenizer.tokenToId("<|transcribe|>");
+    seq.push_back(langToken);
+    seq.push_back(taskToken);
+  }
+
+  seq.push_back(this->timestampBeginToken);
+
+  return seq;
+}
+
+GenerationResult ASR::generate(std::span<const float> waveform,
+                               float temperature,
+                               const DecodingOptions &options) const {
+  std::vector<float> encoderOutput = this->encode(waveform);
+
+  std::vector<int32_t> sequenceIds = this->getInitialSequence(options);
+  const size_t initialSequenceLenght = sequenceIds.size();
+  std::vector<float> scores;
+
+  while (sequenceIds.size() <= ASR::maxDecodeLength) {
+    std::vector<float> logits = this->decode(sequenceIds, encoderOutput);
+    std::vector<float> probs = this->softmaxWithTemperature(
+        logits, (temperature == 0 ? 1.0f : temperature));
```

Comment:
```suggestion
        logits, (temperature == 0.0f ? 1.0f : temperature));
```
And maybe use function for comparing floats with tolerance?